### PR TITLE
feat(net): implement the address resolution network struct functions

### PR DESCRIPTION
**Pull request by Green-Sky**
- [x] implement `getaddrinfo` and `freeaddrinfo`
- [x] do we need stocktype/protocol?
- [x] replace all uses of `getaddrinfo()`
  - [x] `addr_resolve()`
  - [x] `net_getipport()`

---

- [x] depends on https://github.com/TokTok/c-toxcore/pull/2774

---

This pr finally allows us to implement a regression test for https://github.com/TokTok/c-toxcore/pull/2361

Also allows fuzzing returned results from DNS.

---

~This pr implements a different approach to disabling DNS, compared to https://github.com/TokTok/c-toxcore/pull/2694, here we can simply replace the functions pointers in the network object with dummies/stubs.~

- - -
This change is [Reviewable](https://reviewable.io/reviews/TokTok/c-toxcore/2773)

### CODE REVIEW

File: toxcore/network.c [hunk 1]
```diff
@@ -1833,19 +1968,16 @@ static int addr_resolve(const Network *ns, const char *address, IP *to, IP *extr
     const Family tox_family = to->family;
     const int family = make_family(tox_family);
 
-    struct addrinfo hints = {0};
-    hints.ai_family   = family;
-    hints.ai_socktype = SOCK_DGRAM; // type of socket Tox uses.
```
Comment (Green-Sky) [inline toxcore/network.c:1838]: we are looking specifically for udp here, is this even correct??

File: toxcore/net_obj.h
Comment (iphydf) [inline toxcore/net_obj.h:1]: Can you either undo this move or do it in a separate PR?

**Review by iphydf — CHANGES_REQUESTED** [review]

File: toxcore/net_obj.h
Comment (Green-Sky) [reply toxcore/net_obj.h:1]: Yes, I specifically kept this as a separate commit.

**Comment by Green-Sky** [comment]
How about the no dns variant? keep or drop?

File: toxcore/network.c [hunk 2]
```diff
@@ -592,6 +592,108 @@ static int sys_setsockopt(void *obj, Socket sock, int level, int optname, const
...
+
+        // do we need to check socktype/protocol?
+    }
+
+    assert(max_count >= result);
+
+    // HACK: use default memory allocator
+    const Memory* mem = os_memory();
+
+    *addrs = (Network_Addr *)mem_valloc(mem, result, sizeof(Network_Addr));
```
Comment (iphydf) [inline toxcore/network.c:639]: Make a temporary variable and out-param-return at the end.

**Review by iphydf — CHANGES_REQUESTED** [review]

File: toxcore/network.c (hunk 2 above)
Comment (Green-Sky) [reply toxcore/network.c:639]: ... only setting `addrs` on success. I see.

Comment (Green-Sky) [reply toxcore/network.c:639]: should I keep the initialization to null? L601

File: toxcore/network.c [hunk 3]
```diff
@@ -607,8 +711,10 @@ static const Network_Funcs os_network_funcs = {
     sys_socket_nonblock,
     sys_getsockopt,
     sys_setsockopt,
+    sys_getaddrinfo,
+    sys_freeaddrinfo
```
Comment (iphydf) [inline toxcore/network.c:715]: nit: add comma

File: toxcore/network.c [hunk 4]
```diff
@@ -592,6 +592,97 @@ static int sys_setsockopt(void *obj, Socket sock, int level, int optname, const
...
+
+    const int rc = getaddrinfo(address, nullptr, &hints, &infos);
+
+    // Lookup failed.
+    if (rc != 0) {
+        // TODO(Green-Sky): log error
+        return 0;
+    }
+
+    const int32_t max_count = (INT32_MAX) / sizeof(Network_Addr);
```
[the remark this answers is not in the document]
Comment (iphydf) [reply toxcore/network.c:621]: The styler just wants you to remove the redundant `()` around `INT32_MAX`.

[the remark this answers is not in the document]
Comment (Green-Sky) [reply toxcore/network.c:621]: But they fail without it too, just differently.

I think we had this before too, but different.

[the remark this answers is not in the document]
Comment (Green-Sky) [reply toxcore/network.c]: HUH?? It just works now :confused:

**Review by iphydf — APPROVED** [review]

File: toxcore/network.c (hunk 1 above)
Comment (Green-Sky) [reply toxcore/network.c:1849]: It should work either way.
